File: fetch_schedules.mjs
#!/usr/bin/env node
/**
 * fetch_schedules.mjs
 * Fetches GTFS stop_times.txt + trips.txt + calendar.txt from data.gov.my
 * and rebuilds public/station_schedules.json
 * Does NOT touch station_coords.json or rail_tracks.json
 */

import https from "https";
import http from "http";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OUT_FILE = path.join(__dirname, "public", "station_schedules.json");

// Prasarana GTFS base
const BASE = "https://api.data.gov.my/gtfs-static/prasarana";

function fetchText(url) {
  return new Promise((resolve, reject) => {
    console.log(`  GET ${url}`);
    const proto = url.startsWith("https") ? https : http;
    proto.get(url, { timeout: 60000 }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return fetchText(res.headers.location).then(resolve).catch(reject);
      }
      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode} for ${url}`));
      }
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
      res.on("error", reject);
    }).on("error", reject);
  });
}

function parseCSV(text) {
  const lines = text.trim().split(/\r?\n/);
  if (lines.length < 2) return [];
  const headers = lines[0].split(",").map(h => h.trim().replace(/^\uFEFF/, ""));
  return lines.slice(1).map(line => {
    const vals = [];
    let cur = "";
    let inQ = false;
    for (const ch of line) {
      if (ch === '"') { inQ = !inQ; }
      else if (ch === "," && !inQ) { vals.push(cur.trim()); cur = ""; }
      else { cur += ch; }
    }
    vals.push(cur.trim());
    const obj = {};
    headers.forEach((h, i) => { obj[h] = (vals[i] || "").trim(); });
    return obj;
  });
}

// Infer a line ID from route_id + headsign text
function inferLineId(routeId = "", headsign = "") {
  const s = (routeId + " " + headsign).toUpperCase();
  if (/\bKJ\b|KELANA JAYA/.test(s)) return "KJ";
  if (/\bSP\b|SRI PETALING/.test(s)) return "SP";
  if (/\bAG\b|AMPANG/.test(s)) return "AG";
  if (/\bKG\b|KAJANG|PUTRAJAYA MRT/.test(s)) return "KG";
  if (/\bPY\b|PUTRAJAYA SENTRAL|PUTRAJAYA LINE/.test(s)) return "PY";
  if (/\bMR\b|MONORAIL/.test(s)) return "MR";
  if (/\bSB\b|BRT|SUNWAY/.test(s)) return "BRT";
  if (/\bSA\b|SHAH ALAM/.test(s)) return "SA";
  return null;
}

async function main() {
  console.log("Fetching GTFS schedule data from Prasarana...\n");

  // 1. Stops
  console.log("Fetching stops.txt...");
  const stopsText = await fetchText(`${BASE}/stops.txt`);
  const stops = parseCSV(stopsText);
  const stopNameById = {};
  for (const row of stops) {
    if (row.stop_id && row.stop_name) {
      stopNameById[row.stop_id] = row.stop_name.trim();
    }
  }
  console.log(`  -> ${Object.keys(stopNameById).length} stops\n`);

  // 2. Trips
  console.log("Fetching trips.txt...");
  const tripsText = await fetchText(`${BASE}/trips.txt`);
  const trips = parseCSV(tripsText);
  const tripById = {};
  for (const row of trips) {
    tripById[row.trip_id] = {
      route_id: row.route_id || "",
      service_id: row.service_id || "",
      headsign: row.trip_headsign || "",
      direction_id: row.direction_id || "0",
    };
  }
  console.log(`  -> ${Object.keys(tripById).length} trips\n`);

  // 3. Calendar
  console.log("Fetching calendar.txt...");
  const calendarById = {};
  try {
    const calText = await fetchText(`${BASE}/calendar.txt`);
    const calRows = parseCSV(calText);
    for (const row of calRows) {
      const sid = row.service_id;
      if (!sid) continue;
      const weekday = ["monday","tuesday","wednesday","thursday","friday"].some(d => row[d] === "1");
      calendarById[sid] = {
        weekday,
        saturday: row.saturday === "1",
        sunday: row.sunday === "1",
      };
    }
    console.log(`  -> ${Object.keys(calendarById).length} service entries\n`);
  } catch(e) {
    console.warn(`  calendar.txt not available: ${e.message}\n`);
  }

  // 4. Stop times (large file)
  console.log("Fetching stop_times.txt (this may take a while)...");
  const stText = await fetchText(`${BASE}/stop_times.txt`);
  const stRows = parseCSV(stText);
  console.log(`  -> ${stRows.length} stop_time rows\n`);

  // Build: stopId -> tripId -> departureTime (normalized HH:MM)
  const rawStopTimes = {}; // stop_id -> { trip_id: "HH:MM" }
  for (const row of stRows) {
    const sid = row.stop_id;
    const tid = row.trip_id;
    const dep = (row.departure_time || row.arrival_time || "").trim();
    if (!sid || !tid || !dep) continue;
    // Normalize past-midnight times (GTFS allows 25:xx etc.)
    const parts = dep.split(":");
    const h = parseInt(parts[0], 10) % 24;
    const normalized = `${String(h).padStart(2,"0")}:${parts[1]}`;
    if (!rawStopTimes[sid]) rawStopTimes[sid] = {};
    rawStopTimes[sid][tid] = normalized;
  }

  console.log("Building schedules...");
  const schedules = {}; // StationName -> LineId -> dirKey -> { weekday, saturday, sunday }

  for (const [stopId, depsByTrip] of Object.entries(rawStopTimes)) {
    const rawName = stopNameById[stopId];
    if (!rawName) continue;

    // Normalize name: strip trailing qualifiers, uppercase
    const normName = rawName
      .toUpperCase()
      .replace(/\s+STATION\s*$/, "")
      .replace(/\s+(LRT|MRT|MONORAIL)\s*$/, "")
      .trim();

    for (const [tripId, depTime] of Object.entries(depsByTrip)) {
      const trip = tripById[tripId];
      if (!trip) continue;

      const lineId = inferLineId(trip.route_id, trip.headsign);
      if (!lineId) continue;

      // Classify service days
      const svcCal = calendarById[trip.service_id];
      let dayTypes = [];
      if (svcCal) {
        if (svcCal.weekday) dayTypes.push("weekday");
        if (svcCal.saturday) dayTypes.push("saturday");
        if (svcCal.sunday) dayTypes.push("sunday");
      } else {
        // Fallback from service_id name
        const sid = trip.service_id.toLowerCase();
        if (sid.includes("sat")) dayTypes.push("saturday");
        else if (sid.includes("sun")) dayTypes.push("sunday");
        else if (sid.includes("wday") || sid.includes("weekday")) dayTypes.push("weekday");
        else dayTypes = ["weekday", "saturday", "sunday"]; // assume all
      }
      if (dayTypes.length === 0) dayTypes = ["weekday"];

      const dirKey = trip.headsign || `Direction ${trip.direction_id}`;

      if (!schedules[normName]) schedules[normName] = {};
      if (!schedules[normName][lineId]) schedules[normName][lineId] = {};
      if (!schedules[normName][lineId][dirKey]) {
        schedules[normName][lineId][dirKey] = { weekday: [], saturday: [], sunday: [] };
      }
      for (const dt of dayTypes) {
        schedules[normName][lineId][dirKey][dt].push(depTime);
      }
    }
  }

  // Sort and deduplicate
  let stationCount = 0;
  for (const stationData of Object.values(schedules)) {
    stationCount++;
    for (const lineData of Object.values(stationData)) {
      for (const dirData of Object.values(lineData)) {
        for (const dt of ["weekday", "saturday", "sunday"]) {
          dirData[dt] = [...new Set(dirData[dt])].sort();
        }
      }
    }
  }

  fs.writeFileSync(OUT_FILE, JSON.stringify(schedules, null, 2));
  console.log(`\n✓ Wrote ${stationCount} stations to ${OUT_FILE}`);

  // Show sample
  const sample = Object.keys(schedules).slice(0, 3);
  for (const k of sample) {
    console.log(`\n  ${k}:`);
    for (const [lineId, lineData] of Object.entries(schedules[k])) {
      for (const [dirKey, dirData] of Object.entries(lineData)) {
        const wdTimes = dirData.weekday.slice(0, 5);
        console.log(`    [${lineId}] ${dirKey}: ${wdTimes.join(", ")} ... (${dirData.weekday.length} total weekday departures)`);
      }
    }
  }
}

main().catch(err => { console.error(err); process.exit(1); });

File: fetch_schedules.py
#!/usr/bin/env python3
"""
fetch_schedules.py
Fetches the Prasarana GTFS ZIP (rapid-rail-kl) from data.gov.my
and rebuilds public/station_schedules.json with full timetable data.
Does NOT touch station_coords.json or rail_tracks.json.
"""

import urllib.request
import json
import io
import zipfile
import os
import re
from collections import defaultdict

GTFS_URL = "https://api.data.gov.my/gtfs-static/prasarana?category=rapid-rail-kl"
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUT_FILE = os.path.join(SCRIPT_DIR, "public", "station_schedules.json")


def fetch_zip(url):
    print(f"Downloading {url} ...")
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        data = resp.read()
    print(f"  -> Downloaded {len(data):,} bytes")
    return data


def read_csv(zf, filename):
    """Read a CSV from the ZIP into a list of dicts."""
    try:
        with zf.open(filename) as f:
            text = f.read().decode("utf-8-sig")  # handles BOM
        lines = text.strip().splitlines()
        if len(lines) < 2:
            return []
        headers = [h.strip() for h in lines[0].split(",")]
        rows = []
        for line in lines[1:]:
            vals = [v.strip() for v in line.split(",")]
            row = {headers[i]: vals[i] if i < len(vals) else "" for i in range(len(headers))}
            rows.append(row)
        return rows
    except Exception as e:
        print(f"  Warning: could not read {filename}: {e}")
        return []


def infer_line_id(route_id="", headsign=""):
    s = (route_id + " " + headsign).upper()
    if re.search(r'\bKJ\b|KELANA JAYA', s): return "KJ"
    if re.search(r'\bSP\b|SRI PETALING', s): return "SP"
    if re.search(r'\bAG\b|(?<!SRI )AMPANG', s): return "AG"
    if re.search(r'\bKG\b|KAJANG|PUTRAJAYA MRT', s): return "KG"
    if re.search(r'\bPY\b|PUTRAJAYA SENTRAL|PUTRAJAYA LINE', s): return "PY"
    if re.search(r'\bMR\b|MONORAIL', s): return "MR"
    if re.search(r'\bSB\b|BRT|SUNWAY', s): return "BRT"
    if re.search(r'\bSA\b|SHAH ALAM', s): return "SA"
    return None


def normalize_time(t):
    """Normalize GTFS time (may be >24:xx for past-midnight) to HH:MM."""
    parts = t.strip().split(":")
    if len(parts) < 2:
        return None
    h = int(parts[0]) % 24
    m = parts[1].zfill(2)
    return f"{h:02d}:{m}"


def normalize_station_name(name):
    """Remove trailing station qualifiers."""
    name = name.strip().upper()
    name = re.sub(r'\s+STATION\s*$', '', name)
    name = re.sub(r'\s+(LRT|MRT|MONORAIL|KTM)\s*$', '', name)
    return name.strip()


def main():
    # 1. Download ZIP
    zip_data = fetch_zip(GTFS_URL)
    zf = zipfile.ZipFile(io.BytesIO(zip_data))
    print(f"  ZIP contains: {[n for n in zf.namelist() if not n.startswith('__')]}\n")

    # 2. Stops
    print("Parsing stops.txt...")
    stops = read_csv(zf, "stops.txt")
    stop_name_by_id = {}
    for row in stops:
        sid = row.get("stop_id", "")
        sname = row.get("stop_name", "")
        if sid and sname:
            stop_name_by_id[sid] = sname.strip()
    print(f"  -> {len(stop_name_by_id)} stops")

    # 3. Trips
    print("Parsing trips.txt...")
    trips = read_csv(zf, "trips.txt")
    trip_by_id = {}
    for row in trips:
        tid = row.get("trip_id", "")
        if tid:
            trip_by_id[tid] = {
                "route_id": row.get("route_id", ""),
                "service_id": row.get("service_id", ""),
                "headsign": row.get("trip_headsign", ""),
                "direction_id": row.get("direction_id", "0"),
            }
    print(f"  -> {len(trip_by_id)} trips")

    # 4. Calendar
    print("Parsing calendar.txt...")
    calendar = {}
    for row in read_csv(zf, "calendar.txt"):
        sid = row.get("service_id", "")
        if not sid:
            continue
        weekday = any(row.get(d, "0") == "1" for d in ["monday","tuesday","wednesday","thursday","friday"])
        calendar[sid] = {
            "weekday": weekday,
            "saturday": row.get("saturday", "0") == "1",
            "sunday": row.get("sunday", "0") == "1",
        }
    print(f"  -> {len(calendar)} service calendar entries")

    # 5. Stop times
    print("Parsing stop_times.txt (large file)...")
    st_rows = read_csv(zf, "stop_times.txt")
    print(f"  -> {len(st_rows):,} rows")

    # Build stopId -> tripId -> departureTime
    raw_stop_times = defaultdict(dict)  # stop_id -> {trip_id: time}
    for row in st_rows:
        sid = row.get("stop_id", "")
        tid = row.get("trip_id", "")
        dep = row.get("departure_time", "") or row.get("arrival_time", "")
        if not (sid and tid and dep):
            continue
        t = normalize_time(dep)
        if t:
            raw_stop_times[sid][tid] = t

    # 6. Build schedules
    print("\nBuilding schedules...")
    # station_norm_name -> lineId -> dirKey -> {weekday, saturday, sunday: [times]}
    schedules = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: {"weekday": [], "saturday": [], "sunday": []})))

    for stop_id, deps_by_trip in raw_stop_times.items():
        raw_name = stop_name_by_id.get(stop_id)
        if not raw_name:
            continue
        norm_name = normalize_station_name(raw_name)

        for trip_id, dep_time in deps_by_trip.items():
            trip = trip_by_id.get(trip_id)
            if not trip:
                continue
            line_id = infer_line_id(trip["route_id"], trip["headsign"])
            if not line_id:
                continue

            # Day classification
            svc_cal = calendar.get(trip["service_id"])
            day_types = []
            if svc_cal:
                if svc_cal["weekday"]: day_types.append("weekday")
                if svc_cal["saturday"]: day_types.append("saturday")
                if svc_cal["sunday"]: day_types.append("sunday")
            else:
                sid_lower = trip["service_id"].lower()
                if "sat" in sid_lower:
                    day_types = ["saturday"]
                elif "sun" in sid_lower:
                    day_types = ["sunday"]
                else:
                    day_types = ["weekday"]

            if not day_types:
                day_types = ["weekday"]

            dir_key = trip["headsign"] or f"Direction {trip['direction_id']}"

            for dt in day_types:
                schedules[norm_name][line_id][dir_key][dt].append(dep_time)

    # 7. Deduplicate and sort
    result = {}
    for station_name, line_data in sorted(schedules.items()):
        result[station_name] = {}
        for line_id, dir_data in sorted(line_data.items()):
            result[station_name][line_id] = {}
            for dir_key, times_data in sorted(dir_data.items()):
                result[station_name][line_id][dir_key] = {
                    dt: sorted(set(times)) for dt, times in times_data.items()
                }

    # 8. Write
    with open(OUT_FILE, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2)

    print(f"\n✓ Wrote {len(result)} stations to {OUT_FILE}")

    # Sample output
    for station_name in list(result.keys())[:3]:
        print(f"\n  {station_name}:")
        for line_id, dir_data in result[station_name].items():
            for dir_key, times_data in dir_data.items():
                wday = times_data.get("weekday", [])
                print(f"    [{line_id}] {dir_key}: {wday[:5]} ... ({len(wday)} weekday deps)")


if __name__ == "__main__":
    main()
